gap/attr.gi

```diff
@@ -174,6 +174,201 @@ function(D)
+end
+);
+
+BindGlobal("DIGRAPHS_ChromaticNumberLawler",
+function(D)
+  local n, vertices, subset_colours, s, S, i, I, subset_iter, x,
+  mis, subset_mis;
+  n := DigraphNrVertices(D);
+  vertices := List(DigraphVertices(D));
+  # Store all the Maximal Independent Sets, which can later be used for
+  # calculating the maximal independent sets of induced subgraphs.
+  mis := DigraphMaximalIndependentSets(D);
+  # Convert each MIS to a Blist
+  mis := List(mis, x -> BlistList(vertices, x));
+  # Store current best colouring for each subset
+  subset_colours := ListWithIdenticalEntries(2 ^ n, infinity);
+  # Empty set can be colouring with only one colour.
+  subset_colours[1] := 0;
+  # Iterator for blist subsets.
+  subset_iter := ListWithIdenticalEntries(n, [false, true]);
+  subset_iter := IteratorOfCartesianProduct2(subset_iter);
+  # Skip the first one, which should be the empty set.
+  S := NextIterator(subset_iter);
+  # Iterate over all vertex subsets.
+  for S in subset_iter do
+    # Cartesian iterator is ascending lexicographically, but we want reverse
+    # lexicographic ordering. We treat this as iterating over the complement.
+    # Index the current subset that is being iterated over.
+    s := 1;
+    for x in [1 .. n] do
+      # Need to negate, as we are iterating over the complement.
+      if not S[x] then
+        s := s + 2 ^ (x - 1);
+      fi;
+    od;
+    # Iterate over the maximal independent sets of V[S]
+    subset_mis := DIGRAPHS_MaximalIndependentSetsSubtractedSet(mis, S, infinity);
+    # Flip the list, as we now need the actual set.
+    FlipBlist(S);
+    for I in subset_mis do
+        # Calculate S \ I. This is destructive, but is undone.
+        SubtractBlist(S, I);
+        # Index S \ I
+        i := 1;
+        for x in [1 .. n] do
+          if S[x] then
+            i := i + 2 ^ (x - 1);
+          fi;
+        od;
+        # The chromatic number of this subset is the minimum value of all
+        # the maximal independent subsets of D[S].
+        subset_colours[s] := Minimum(subset_colours[s], subset_colours[i] + 1);
+        # Undo the changes to the subset.
+        UniteBlist(S, I);
+    od;
+  od;
+  return subset_colours[2 ^ n];
+end
+);
+
+BindGlobal("DIGRAPHS_UnderThreeColourable",
+function(D)
+  local nr;
+  nr := DigraphNrVertices(D);
+  if DigraphHasLoops(D) then
+    ErrorNoReturn("the argument <D> must be a digraph with no loops,");
+  elif nr = 0 then
+    return 0;  # chromatic number = 0 iff <D> has 0 verts
+  elif IsNullDigraph(D) then
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  elif DigraphColouring(D, 3) <> fail then
+               # Check if there is a 3 colouring
+    return 3;
+  fi;
+  return infinity;
+end
+);
+
+BindGlobal("DIGRAPHS_ChromaticNumberByskov",
+function(D, Byskov)
```

Comment:
Thanks a lot for making these changes @EwanGilligan. I've spotted another thing that needs to be changed.
```suggestion
function(D)
```
I'm not sure why the CI is crashing so much... Perhaps making that change will magically fix it?!

Reply:
Nope.

Reply:
Ah that seems to be the issue, was searching for why everything was crashing. Thanks @wilfwilson !

Reply:
Unfortunately the situation persists on my computer, and probably will do on Azure/GitHub Actions (when those jobs finally run).

However, I can make the problem disappear by commenting out the lines:
```
  if ValueOption("lawler") <> fail then
    return DIGRAPHS_ChromaticNumberLawler(D);
  elif ValueOption("byskov") <> fail then
    return DIGRAPHS_ChromaticNumberByskov(D);
  fi;
```
(i.e. if your code is bypassed).

Now, this doesn't prove that your code is buggy, but it at least strongly suggests that your code is at least running another big of code which is buggy. And given the kind of crash that's happening, this suggests that it's whichever code your calling that is written in C.

For me, that points to `DigraphMaximalIndependentSets`...

Reply:
I'm poking around a bit more.

Reply:
I don't believe anything should have changed, such that it would now run out of memory. Checking back it passed all CI previously, before the new method selection was used. I'll investigate to see if I can reproduce it, but it all works fine on my end. Interestingly (of the ones that have currently run) only the github action for 4.11 seems to fail, with the 4.10 one passing